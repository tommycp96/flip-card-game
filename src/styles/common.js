import { Dimensions, StyleSheet } from 'react-native';

const dim = Dimensions.get('window');
const deviceWidth = dim.width;
const deviceHeight = dim.height;

export const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#e5e5e5',
    width: deviceWidth,
    borderWidth: 1,
    margin: 'auto'
  },
  gridOutline: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    width: '100%',
    paddingHorizontal: 20,
    maxWidth: 600,
    justifyContent: 'space-between'
  },
  scoreContainer: {
    marginTop: 8,
    flexDirection: 'row',
    width: '100%',
    justifyContent: 'space-evenly'
  },
  steps: {
    fontSize: 20
  },
  matches: {
    fontSize: 20
  },
  button: {
    backgroundColor: '#4fc3f7',
    padding: 8,
    marginVertical: 8,
    borderRadius: 8,
    justifyContent: 'center',
    alignItems: 'center',
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 4
    },
    shadowOpacity: 0.22,
    shadowRadius: 2.22,
    elevation: 4
  }
});
