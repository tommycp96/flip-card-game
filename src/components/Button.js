import React from 'react';
import { TouchableOpacity, Text } from 'react-native';
import { styles } from '../styles/common';

const Button = ({ title, onPress }) => {
  return (
    <TouchableOpacity style={styles.button} onPress={onPress}>
      <Text style={{ fontSize: 14, fontWeight: 'bold' }}>{title}</Text>
    </TouchableOpacity>
  );
};

export default Button;
