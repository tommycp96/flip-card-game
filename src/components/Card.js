import React, { useRef } from 'react';
import { StyleSheet, Text, TouchableOpacity, Animated } from 'react-native';
import { colors } from '../styles/common';

const Card = ({ char, onPress, index, isMatched, isShow }) => {
  const flipAnimation = useRef(new Animated.Value(0)).current;
  let flipRotation = 0;
  flipAnimation.addListener(({ value }) => (flipRotation = value));

  const flipToFrontStyle = {
    transform: [
      {
        rotateY: flipAnimation.interpolate({
          inputRange: [0, 180],
          outputRange: ['0deg', '180deg']
        })
      }
    ]
  };

  const flipToBackStyle = {
    transform: [
      {
        rotateY: flipAnimation.interpolate({
          inputRange: [0, 180],
          outputRange: ['180deg', '360deg']
        })
      }
    ]
  };

  const flipToFront = () => {
    Animated.timing(flipAnimation, {
      toValue: 180,
      duration: 300,
      useNativeDriver: true
    }).start();
  };

  const flipToBack = () => {
    Animated.timing(flipAnimation, {
      toValue: 0,
      duration: 300,
      useNativeDriver: true
    }).start();
  };

  const flipCard = () => {
    if (!isMatched) {
      flipAnimation.setValue(0);
      !!flipRotation ? flipToBack() : flipToFront();
    }
  };

  const handlePress = () => {
    if (isMatched) return;
    onPress(index);
    flipCard();
  };

  const cardTextValue = () => {
    if (isMatched) return <Text style={styles.matched}>{char}</Text>;
    else if (isShow) return <Text style={styles.shown}>{char}</Text>;
    else return <Text style={styles.hidden}>?</Text>;
  };

  return (
    <TouchableOpacity style={styles.container} onPress={handlePress}>
      <Animated.View
        style={{
          ...styles.card,
          ...flipToFrontStyle,
          backgroundColor:
            !isMatched && !isShow ? colors.primary : colors.secondary
        }}
      >
        {cardTextValue()}
      </Animated.View>
      <Animated.View
        style={{
          ...styles.card,
          ...styles.cardBack,
          ...flipToBackStyle,
          backgroundColor:
            !isMatched && !isShow ? colors.primary : colors.secondary
        }}
      >
        {cardTextValue()}
      </Animated.View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  container: {
    width: '24%',
    minHeight: 100,
    alignItems: 'center',
    justifyContent: 'center'
  },
  card: {
    width: '100%',
    padding: 8,
    marginBottom: 8,
    borderRadius: 8,
    justifyContent: 'center',
    alignItems: 'center',
    minHeight: 100,
    shadowColor: '#000',
    shadowOffset: {
      width: 2,
      height: 4
    },
    shadowOpacity: 0.22,
    shadowRadius: 2.22,
    elevation: 8,
    backfaceVisibility: 'hidden'
  },
  cardBack: {
    position: 'absolute',
    top: 0
  },
  matched: {
    color: 'lightgrey',
    fontSize: 20
  },
  shown: {
    fontWeight: 'bold',
    fontSize: 20
  },
  hidden: {
    fontSize: 20
  }
});

export default Card;
