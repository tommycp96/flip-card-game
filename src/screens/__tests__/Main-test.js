import { View, Text, Alert } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';
import React from 'react';
import Card from '../../components/Card';
import Button from '../../components/Button';
import { styles } from '../../styles/common';
import renderer from 'react-test-renderer';

it('should renders correctly', () => {
  const attempts = 18;
  const matches = 8;
  const tree = renderer
    .create(
      <SafeAreaView style={styles.container}>
        <View style={styles.gridOutline}>
          <Card />
        </View>

        <View style={styles.scoreContainer}>
          <Text style={styles.steps}>Steps: {attempts}</Text>
          <Text style={styles.matches}>Matches: {matches}</Text>
        </View>
        <Button title="Restart" />
      </SafeAreaView>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('should renders alert correctly when isGameWon is true', () => {
  const isGameWon = true;
  const attempts = 27;
  const tree =
    isGameWon &&
    renderer
      .create(
        Alert.alert(
          'Congratulations!',
          `You win this game by ${attempts} steps!`,
          [{ text: 'Try another round', onPress: () => initialiseGame() }]
        )
      )
      .toJSON();
  expect(tree).toMatchSnapshot();
});
