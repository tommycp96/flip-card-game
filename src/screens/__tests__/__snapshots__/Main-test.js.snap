// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should renders alert correctly when isGameWon is true 1`] = `null`;

exports[`should renders correctly 1`] = `
<RNCSafeAreaView
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#e5e5e5",
      "borderWidth": 1,
      "flex": 1,
      "justifyContent": "center",
      "margin": "auto",
      "width": 750,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "flexWrap": "wrap",
        "justifyContent": "space-between",
        "maxWidth": 600,
        "paddingHorizontal": 20,
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "minHeight": 100,
          "opacity": 1,
          "width": "24%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#4fc3f7",
            "borderRadius": 8,
            "elevation": 4,
            "justifyContent": "center",
            "marginBottom": 8,
            "minHeight": 100,
            "padding": 8,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 4,
              "width": 0,
            },
            "shadowOpacity": 0.22,
            "shadowRadius": 2.22,
            "transform": Array [
              Object {
                "rotateY": "0deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
            }
          }
        >
          ?
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backfaceVisibility": "hidden",
            "backgroundColor": "#4fc3f7",
            "borderRadius": 8,
            "elevation": 4,
            "justifyContent": "center",
            "marginBottom": 8,
            "minHeight": 100,
            "padding": 8,
            "position": "absolute",
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 4,
              "width": 0,
            },
            "shadowOpacity": 0.22,
            "shadowRadius": 2.22,
            "top": 0,
            "transform": Array [
              Object {
                "rotateY": "180deg",
              },
            ],
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
            }
          }
        >
          ?
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-evenly",
        "marginTop": 8,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 20,
        }
      }
    >
      Steps: 
      18
    </Text>
    <Text
      style={
        Object {
          "fontSize": 20,
        }
      }
    >
      Matches: 
      8
    </Text>
  </View>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#4fc3f7",
        "borderRadius": 8,
        "elevation": 4,
        "justifyContent": "center",
        "marginVertical": 8,
        "opacity": 1,
        "padding": 8,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.22,
        "shadowRadius": 2.22,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 20,
        }
      }
    >
      Restart
    </Text>
  </View>
</RNCSafeAreaView>
`;
